src/jdk.jlink/share/classes/jdk/tools/jlink/internal/TaskHelper.java

```diff
@@ -419,6 +420,77 @@ private PluginsConfiguration getPluginsConfig(Path output, Map<String, String> l
+                        systemModulesPlugin = p;
+                    }
+                    if (p instanceof AddOptionsPlugin) {
+                        addOptionsPlugin = p;
+                    }
+                }
+                // Certain system module classes get generated and SystemModulesMap replaced by the
+                // SystemModulesPlugin. Filter previously generated classes so the
+                // set of classes match the set of the packaged modules link.
+                String systemModulesPattern = "regex:/java\\.base/jdk/internal/module/SystemModules\\$.*\\.class";
+                // The default OpenJDK build generates JLI Species classes so they
+                // would be in the run-time image we derive from if we don't filter
+                // them. This filter corresponds to the set of classes part of the
+                // jmods.
+                String speciesPattern = "regex:/java\\.base/java/lang/invoke/BoundMethodHandle\\$Species_(?:D|DL|I|IL|LJ|LL).*\\.class";
+                // The AddResourcePlugin (used by AddOptionsPlugin)
+                String addOptionsGlob = "glob:" + AddOptionsPlugin.OPTS_FILE;
+                String saveJlinkOptsGlob = "glob:/jdk.jlink/" + JlinkTask.OPTIONS_RESOURCE;
+                String additionalPatterns = systemModulesPattern + "," +
+                                            speciesPattern + "," +
+                                            addOptionsGlob + "," +
+                                            saveJlinkOptsGlob;
+                List<Map<String, String>> excludeResConfig = null;
+                if (excludeResourcePlugin == null) {
+                    // no existing 'exclude-resources' setting
+                    excludeResourcePlugin = PluginRepository.getPlugin("exclude-resources", ModuleLayer.boot());
+                    excludeResConfig = new ArrayList<>();
+                    excludeResConfig.add(Map.of("exclude-resources", additionalPatterns));
+                    pluginToMaps.put(excludeResourcePlugin, excludeResConfig);
+                } else {
+                    excludeResConfig = pluginToMaps.get(excludeResourcePlugin);
+                    // currently last exclude-resources wins
+                    Map<String, String> lastConfig = excludeResConfig.get(excludeResConfig.size() - 1);
+                    String existingPattern = lastConfig.get("exclude-resources");
+                    lastConfig.put("exclude-resources", existingPattern + "," + additionalPatterns);
+                    excludeResConfig.set(excludeResConfig.size() - 1, lastConfig);
+                }
+                // If the system modules plug-in is disabled, we add the
+                // -Djdk.system.module.finder.disableFastPath property as the
+                // SystemModulesMap class isn't guaranteed to be correct for the
+                // current module set.
+                if (systemModulesPlugin == null) {
```

Comment:
Suggest to fail if system modules plugin does not exist which should not happen.  The `disableFastPath` system property was added for testing only.